Support Pagination and Column Blacklist

twodayslate opened the pull request:
Used [black](https://black.readthedocs.io/en/stable/index.html) for code formatting

For the [sqla.py](examples/sqla.py) example you can try the following:

http://localhost:8888/api/posts/?limit=1
http://localhost:8888/api/posts/?limit=2&page=2

You'll see the server has a [hard limit of 12](https://github.com/lockie/tornado_jsonapi/pull/2/files#diff-17ecd233edf87877621cce5904a6f237R33) but you can request a smaller limit if desired.

Blacklisting is as simple as [adding the column](https://github.com/lockie/tornado_jsonapi/pull/2/files#diff-17ecd233edf87877621cce5904a6f237R48) to the resource blacklist

twodayslate committed:
Supports pagenation for sqla
black the code

twodayslate committed:
Added column blacklist

twodayslate committed:
black formatting

twodayslate committed:
Bug fix for jsonapi_limit setting

twodayslate committed:
Supports async databases

lockie commented on examples/async.py:
```diff
@@ -89,9 +89,13 @@ def delete(self, id_):
         return [Posts.ResourceObject(self, p) for p in self.data]
 
+    @gen.coroutine
+    def list_count_(self):
```
I suggest naming it `list_count` or simply `count`, because there is no name conflict with built-in function, as there is with `list`.

lockie commented on tornado_jsonapi/__init__.py:
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python3
 # vim: set fileencoding=utf8 :
 
-'''
```
Let me guess, you're using black formatter? :sweat_smile:

lockie replied:
Oh, yeah, I haven't read your changes description :sweat_smile:

lockie commented on tornado_jsonapi/handlers.py:
```diff
@@ -17,194 +17,248 @@
+        server_limit = (
+            self.settings["jsonapi_limit"] if "jsonapi_limit" in self.settings else 0
+        )
+        print("server limit", server_limit)
```
I suggest removing any debug `print`ing or using `logging` for that purpose (if you really need it).

lockie commented on tornado_jsonapi/handlers.py:
```diff
@@ -17,194 +17,248 @@
+            self.render(
+                res,
+                additional={
+                    "list_limits": {"total": count, "limit": limit, "page": page,}
```
Why not name it simply `limits`?

lockie commented on tornado_jsonapi/resource.py:
```diff
@@ -187,23 +224,23 @@ def __enter__(self):
     return wrapper(pool, connection, transaction)
 
 
 class DBAPI2Resource(Resource):
```
I guess for your changes to be consistent, you'd also have to append blacklist and paging support to `DBAPI2Resource`.

twodayslate replied:
Complete. I am not sure if limit #,# is supported for all platforms.

lockie reviewed (changes requested):
Thanks for your efforts! There's some minor naming issues, an extra debug print and a need to modify `DBAPI2Resource` as you did with `SQLAlchemyResource`. Otherwise it looks good to me.

lockie commented on examples/sqla.py:
```diff
@@ -29,17 +30,28 @@ def main():
+    s.commit()
+
+    postResource = tornado_jsonapi.resource.SQLAlchemyResource(Post, Session)
+    postResource.blacklist.append(Post.hideMe)
```
Maybe it would be better to use simple strings as field designators here.

twodayslate replied:
Strangely enough both work. I have added a strings example to sqla.py

lockie commented on tornado_jsonapi/resource.py:
```diff
@@ -315,33 +364,49 @@ def update(self, id_, attributes):
+        count = self.list_()
+        while is_future(count):
+            count = yield count
+        return len(count)
```
This method looks quite ineffective. Perhaps it would be better to issue `select count(1)` SQL statement.

twodayslate replied:
Let me know if I implemented this correctly

twodayslate committed:
Renamed list_count_ to just list_count

twodayslate committed:
Removed extra print and renamed list_limits to just limits as suggested

twodayslate committed:
Added extra blacklist example using strings

twodayslate committed:
Using select count instead of len(count) for dbapi

twodayslate committed:
Added limits to DBAPI

twodayslate commented:
Thank you for the indepth review.

lockie commented:
You're welcome!
There is minor issue with SQL limit/offset, and then I think your PR is ready to merge.
There are also some tests failing, but I'll fix those myself after.

twodayslate committed:
using limits and offsets and formatting

twodayslate commented:
Does that fix the issue(s)?

edit://

I changed limit and page to page[size] and page[number] respectfully to match [JSON:API](https://jsonapi.org/format/#fetching-pagination).

I have modified the meta as that is where the limit stuff is supposed to go based on [JSON:API](https://jsonapi.org/format/#document-meta). I also added a [links sections](https://jsonapi.org/format/#document-links) based on [JSON:API pagination](https://jsonapi.org/format/#fetching-pagination) information as well. 

Let me know if you want any other changes.

I added an auth.py example as I don't think it is obvious to new users that authentication can be added to the API (and perhaps should be). 

I added a spec.py example which should help with the automatic documentation generation TODO you have. You can use [sphinx-jsonschema](https://sphinx-jsonschema.readthedocs.io/en/latest/directive.html#an-example) to generate documentation. Anything else and I think this project would become very bloated. Should this be an official handle?

edit2://

StructuralWalker was bugging out for me with relationships for SQLAlchemy so I now use NoForeignKeyWalker and support relationships when rendering. Creating or updating not currently tested

lockie commented:
Yeah, 964f72527cc7775b0c8d764939060450358a4505 fixes the limit/offset issue.
For other changes, you most definitely should've open new pull request with new branch, so that we can eat an elephant one bite at a time. Do you mind doing it and removing those new commits from this PR? Use git force push if you must.

lockie commented:
Thanks again for your efforts!
